### ISSUE TITLE
ERROR -- : TaskTimeout at start of supervised instance of ...

### ISSUE THREAD

**Issue by felixalias** [celluloid/celluloid#680]
Hi, I'm trying to use Celluloid v0.17.1.2's Supervision::Container with JRuby 1.7.11. 

The supervision class looks something like this (snipped down):
```ruby
class Manager < Celluloid::Supervision::Container
  def initialize
    super
    set_supervisors
  end

  def set_supervisors
    supervise CrashyWorker # in reality, loaded from a config file
  end
end


module Worker
  class CrashyWorker
    include Celluloid

    def initialize params = {}
      puts "Started #{self}"
      async.work
    end

    def work
      loop do
        puts "Entering \#crash_forever"
        sleep 0.1
        raise "error!"
      end
    end

  end
end


# Elsewhere..
loop do
      @manager = Worker::Manager.run!
      sleep 5 while @manager.alive?
      logger.error "Worker::Manager crashed, restarting"
end
```

After Celluloid restarts it a fair number of times successfully (hundreds, maybe?), it fails and the whole app crashes:
```
E, [2015-09-22T14:44:53.452000 #11255] ERROR -- : TaskTimeout at start of supervised instance of Worker::CrashyWorker
E, [2015-09-22T14:44:53.460000 #11255] ERROR -- : Actor crashed!
Celluloid::TaskTimeout: linking timeout of 30 seconds exceeded with receiver: #<Worker::CrashyWorker:0x2df881c6>
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/actor.rb:208:in `linking_request'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/task.rb:93:in `exclusive'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid.rb:421:in `exclusive'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/actor.rb:182:in `linking_request'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/actor.rb:56:in `monitor'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/actor.rb:67:in `link'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid.rb:206:in `new_link'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-supervision-0.20.1.1/lib/celluloid/supervision/container/instance.rb:34:in `start'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-supervision-0.20.1.1/lib/celluloid/supervision/container/instance.rb:57:in `restart'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-supervision-0.20.1.1/lib/celluloid/supervision/container.rb:121:in `restart_actor'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/task.rb:97:in `exclusive'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/task.rb:95:in `exclusive'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid.rb:421:in `exclusive'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-supervision-0.20.1.1/lib/celluloid/supervision/container.rb:121:in `restart_actor'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/cell.rb:15:in `call'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/cell.rb:76:in `task'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/actor.rb:339:in `task'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/task.rb:44:in `initialize'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/task.rb:34:in `initialize'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/task/threaded.rb:23:in `create'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/task/threaded.rb:19:in `create'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-essentials-0.20.2.1/lib/celluloid/internals/thread_handle.rb:14:in `initialize'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-essentials-0.20.2.1/lib/celluloid/internals/thread_handle.rb:13:in `initialize'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/actor/system.rb:78:in `get_thread'
        org/jruby/RubyProc.java:271:in `call'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/group/spawner.rb:53:in `instantiate'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/group/spawner.rb:52:in `instantiate'


22 Sep 2015 14:44:56,309 [ERROR]  (main) Object: Worker::Manager crashed, restarting # this is my debug line


E, [2015-09-22T14:44:56.334000 #11255] ERROR -- : TaskTimeout at start of supervised instance of Worker::CrashyWorker
rake aborted!
linking timeout of 30 seconds exceeded with receiver: #<Worker::CrashyWorker:0x29ca47fa>
lib/worker/manager.rb:36:in `set_supervisors'
org/jruby/RubyArray.java:1613:in `each'
lib/worker/manager.rb:35:in `set_supervisors'
lib/worker/manager.rb:27:in `initialize'
org/jruby/RubyKernel.java:1965:in `public_send'
org/jruby/RubyProc.java:271:in `call'
(celluloid):0:in `remote procedure call'
lib/tasks/worker.rake:12:in `(root)'
lib/tasks/worker.rake:10:in `loop'
lib/tasks/worker.rake:10:in `(root)'
lib/tasks/worker.rake:32:in `(root)'
lib/tasks/worker.rake:34:in `(root)'
Tasks: TOP => worker
(See full trace by running task with --trace)
E, [2015-09-22T14:44:56.346000 #11255] ERROR -- : Actor crashed!
Celluloid::TaskTimeout: linking timeout of 30 seconds exceeded with receiver: #<Worker::CrashyWorker:0x29ca47fa>
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/actor.rb:208:in `linking_request'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/task.rb:97:in `exclusive'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/task.rb:95:in `exclusive'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid.rb:421:in `exclusive'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/actor.rb:182:in `linking_request'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/actor.rb:56:in `monitor'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid/actor.rb:67:in `link'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-0.17.1.2/lib/celluloid.rb:206:in `new_link'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-supervision-0.20.1.1/lib/celluloid/supervision/container/instance.rb:34:in `start'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-supervision-0.20.1.1/lib/celluloid/supervision/container/instance.rb:29:in `initialize'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-supervision-0.20.1.1/lib/celluloid/supervision/container.rb:73:in `add'
        /workplace/felixjo//env/-1.0/test-runtime/jruby1.7.x/ruby/gems/shared/gems/celluloid-supervision-0.20.1.1/lib/celluloid/supervision/deprecate/supervise.rb:96:in `supervise'
        lib/worker/manager.rb:36:in `set_supervisors'
```

As you can see, I've tried adjusting the linking timeout to 30 seconds, as well as setting the task_class to be Celluloid::TaskThread (as well as bumping from 0.16 to 0.17x based on a PR I saw about memory leaks). In a JRuby REPL in my environment, I'm able to spawn as many threads as I practically want to. I'm sort of at a loss of how to debug this further. 

If the way I'm doing things is unsupported and bad (I'm not sure, I've just started using Celluloid), I'm trying to spin off a bunch of distinct long-running workers that I want restarted if they fail.

***Update***:
Using a [simple repro gist](https://gist.github.com/felixalias/a89f59539f9c74695320), it was narrowed down to JVM arguments. Having JRuby 1.7.11/JDK8's --server mode enabled by itself, the case reprod. With --debug, no issue occurred. With InvokeDynamic disabled, the problem went away.

If you're seeing this, you can fix this by adding this to your JVM arguments:
```
-Xcompile.invokedynamic=false
```
Presently, this shouldn't be such a big deal - apparently InvokeDynamic isn't quite up to JDK7 performance anyways (yet).

**Comment by digitalextremist** [comment]
Yes, this can be done differently.

The `supervise` lines ought to happen directly in the class body for your container.

Have you tried that?

**Comment by felixalias** [comment]
They aren't in the class body because they're loaded dynamically from configuration (this class has no knowledge of what workers it will be supervising).

I just tried with the supervise statement in the class body and got the same result.

**Comment by digitalextremist** [comment]
Well, then two things... you ought to try using a configuration object, then deploy the configuration... but if it also happens with `supervise` in the class body... then the actors you are instantiating and supervising likely have issues in their `initialize` methods.

**Comment by felixalias** [comment]
The initialize method really is just as simple as I listed (it's an async call to a method that crashes a lot). Is that an unsupported pattern? If so, what's a better way to ensure that a long-running method stays running?

**Comment by digitalextremist** [comment]
Do any of the initializers create UUID's by chance?

**Comment by felixalias** [comment]
Nope, it's truly as simple as 

```ruby
def initialize
  puts "Starting #{self}"
  async.work
end

def work
   sleep 0.1
   raise
end
```

Does the async call create UUIDs?

**Comment by digitalextremist** [comment]
No.

Can you rerun your test with the `supervise` commands in the class body?

**Comment by felixalias** [comment]
Yes, I ran it without any of the dynamic loading logic as above and got the identical failure scenario.

**Comment by digitalextremist** [comment]
Please post a gist with a runable repro and I will look into this.

**Comment by felixalias** [comment]
Just working on a repro, I have a tiny script that does it. One thing I've noticed right away is that the issue does NOT occur with --debug, but starts happening with --server (JVM arguments).

Further, it goes away with InvokeDynamic disabled (-Xcompile.invokedynamic=false). It's also worth mentioning I'm on JDK8 (which supposedly resolved those issues with InvokeDynamic).. Looks like this is more of a JRuby issue than anything else.

To summarize, problem only occurs with the --server flag AND invokedynamic enabled. I can keep --server, but I need to disable invokedynamic.

Here's my gist:
https://gist.github.com/felixalias/a89f59539f9c74695320

**Comment by digitalextremist** [comment]
For future benefit, can you take your last comment and update your issue description with that excellent information? Thanks for helping to diagnose this. I will look into it.

**Comment by felixalias** [comment]
Sure! Updated, thanks for debugging this with me.

**Comment by headius** [comment]
I concur that this looks like a JRuby issue, and probably one that has been fixed (JRuby 1.7.11 is rather old now). I would strongly recommend trying a more recent version of JRuby, and if you can still reproduce it please report an issue in our tracker. Thank you!

**Comment by digitalextremist** [comment]
Thank you @headius. Closing as we just got it straight from the source. Reopen if you need to!
